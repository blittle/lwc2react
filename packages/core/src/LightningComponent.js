import React from 'react';
import ObservableMembrane from 'observable-membrane';

export default class LightningComponent extends React.Component {
  constructor(template) {
    super();

    const membrane = new ObservableMembrane({
      valueMutated: () => {
        if (this.mounted) this.forceUpdate();
      },
    });

    if (!this.__s) {
      this.__s = membrane.getProxy(Object.assign({}, this.props));
    }

    this.template = React.createRef();

    template.customEvents.forEach((event) => {
      const ref = event[2];
      this[ref] = React.createRef();
    });
  }

  componentDidUpdate(prevProps) {
    this.updateMembrane(prevProps);
  }

  componentDidMount(template) {
    this.mounted = true;
    this.updateMembrane();
    this.addLocalStyles(template);
    this.addEventHandlers(template);
  }

  componentWillUnmount(template) {
    this.mounted = false;
    this.cleanupLocalStyles(template);
    this.cleanupEventHandlers(template);
  }

  addLocalStyles(template) {
    this.stylesheets = [];
    template.stylesheets.forEach((stylesheet) => {
      const sheet = document.createElement('style');
      sheet.type = 'text/css';
      sheet.textContent = stylesheet(
        '[' + template.stylesheetTokens.hostAttribute.toLowerCase() + ']',
        '[' + template.stylesheetTokens.shadowAttribute.toLowerCase() + ']',
        null
      );
      document.head.appendChild(sheet);
      this.stylesheets.push(sheet);
    });
  }

  addEventHandlers(template) {
    template.customEvents.forEach((event) => {
      const name = event[0];
      const cb = event[1];
      const ref = event[2];
      this[cb] = this[cb].bind(this);
      if (ref !== 'template') {
        this[ref].current.template.current.addEventListener(name, this[cb]);
      } else {
        this[ref].current.addEventListener(name, this[cb]);
      }
    });
  }

  cleanupLocalStyles() {
    this.stylesheets.forEach((sheet) => {
      if (sheet.parentNode) sheet.parentNode.removeChild(sheet);
    });
  }

  cleanupEventHandlers(template) {
    template.customEvents.forEach((event) => {
      const name = event[0];
      const cb = event[1];
      const ref = event[3];

      if (ref !== 'template') {
        this[ref].current.template.current.removeEventListener(name, this[cb]);
      } else {
        this[ref].current.removeEventListener(name, this[cb]);
      }
    });
  }

  updateMembrane(prevProps = {}) {
    Object.keys(this.props).forEach((prop) => {
      if (prevProps[prop] !== this.props[prop])
        this.__s[prop] = this.props[prop];
    });
  }
}
