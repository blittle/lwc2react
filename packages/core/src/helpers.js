import { parse, types } from 'recast';

// const b = types.builders;
const n = types.namedTypes;

export function removeLWCCode(ast) {
  const lwcIndex1 = ast.program.body.findIndex(
    (el) => n.ImportDeclaration.check(el) && el.source.value === 'lwc'
  );
  ast.program.body[lwcIndex1] = parse(
    'import React from "react"'
  ).program.body[0];

  let lwcIndex2 = ast.program.body.findIndex(
    (el) => n.ImportDeclaration.check(el) && el.source.value === 'lwc'
  );

  while (lwcIndex2 !== -1) {
    ast.program.body.splice(lwcIndex2, 1);
    lwcIndex2 = ast.program.body.findIndex(
      (el) => n.ImportDeclaration.check(el) && el.source.value === 'lwc'
    );
  }

  const registerDecoratorsIndex = ast.program.body.findIndex(
    (el) =>
      n.ExpressionStatement.check(el) &&
      n.CallExpression.check(el.expression) &&
      el.expression.callee.name === '_registerDecorators'
  );

  if (registerDecoratorsIndex !== -1) {
    ast.program.body.splice(registerDecoratorsIndex, 1);
  }
}
