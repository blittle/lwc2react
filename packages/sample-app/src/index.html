<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <title></title>
    <link rel="stylesheet" href="/css/global.css" />
    <link rel="stylesheet" href="https://blittle-developer-edition.ap1.stmpb.stm.force.com/assets/styles/salesforce-lightning-design-system.min.css?25def0b3bc" />
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="dist/index.js"></script>
  </body>
</html>
