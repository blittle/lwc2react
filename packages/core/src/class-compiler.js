import { parse, types, prettyPrint } from 'recast';
import { removeLWCCode } from './helpers';

const b = types.builders;
const n = types.namedTypes;

function getClassAST(ast) {
  const exportStatement = ast.program.body.find((el) =>
    n.ExportDefaultDeclaration.check(el)
  );
  const exportName = (exportStatement?.declaration?.arguments || [])[0]?.name;

  return (
    ast.program.body.find(
      (element) =>
        n.ClassDeclaration.check(element) &&
        element.superClass.name === 'LightningElement'
    ) ||
    ast.program.body.find(
      (element) =>
        n.ClassDeclaration.check(element) && element.id.name === exportName
    )
  );
}

function getPublicProps(ast) {
  const decoratorsCall = ast.program.body.find(
    (el) =>
      n.ExpressionStatement.check(el) &&
      n.CallExpression.check(el.expression) &&
      el.expression.callee.name === '_registerDecorators'
  );
  const publicProps = decoratorsCall?.expression?.arguments[1]?.properties?.find(
    (prop) => prop.key.name === 'publicProps'
  );
  return publicProps?.value?.properties?.map((prop) => prop.key.name) ?? [];
}

function getClassMethods(classAST) {
  return classAST.body.body
    .filter((el) => n.MethodDefinition.check(el) && el.kind === 'method')
    .map((method) => method.key.name);
}

function convertClass(classAST, publicProps, templateIdentifier) {
  const classMethods = getClassMethods(classAST);

  if (classAST.superClass.name === 'LightningElement') {
    classAST.superClass = b.identifier('LightningComponent');
  }

  convertMethods(classAST, publicProps, classMethods);
  buildRenderMethod(classAST, templateIdentifier);
  buildStyles(classAST, templateIdentifier);
  convertConstructorBlock(
    classAST,
    publicProps,
    classMethods,
    templateIdentifier
  );
  convertLifeCycleMethods(classAST);
}

function convertConstructorBlock(
  classAST,
  publicProps,
  classMethods,
  templateIdentifier
) {
  const constructorAST = classAST.body.body.find(
    (prop) => prop.kind === 'constructor'
  );

  if (constructorAST) {
    constructorAST.value.body = processBlock(constructorAST.value.body, {
      inConstructor: true,
      publicProps,
      classMethods,
    });

    constructorAST.value.body.body[0].expression.arguments = [
      b.identifier(templateIdentifier),
    ];
  } else {
    const a = parse(`
      class A {
        constructor() {
          super(${templateIdentifier});
        }
      }
    `).program.body[0].body.body[0];

    classAST.body.body.unshift(a);
  }
}

function buildRenderMethod(classAST, templateIdentifier) {
  // maybe extending `this` like this is not okay?
  const render = parse(`function render() {
    const prox = Object.assign({}, this, this.__s);
    prox.__proto__ = this.__proto__;
    return ${templateIdentifier}(prox);
  }`);
  classAST.body.body.push(
    b.methodDefinition(
      'method',
      b.identifier('render'),
      b.functionExpression(
        b.identifier('render'),
        [],
        render.program.body[0].body
      )
    )
  );
}

function getMethodIndex(classAST, name) {
  return classAST.body.body.findIndex(
    (el) => n.MethodDefinition.check(el) && el.key.name === name
  );
}

function convertLifeCycleMethods(classAST) {
  const componentDidMount = classAST.body.body.find(
    (el) => n.MethodDefinition.check(el) && el.key.name === 'componentDidMount'
  );
  const componentWillUnmount = classAST.body.body.find(
    (el) =>
      n.MethodDefinition.check(el) && el.key.name === 'componentWillUnmount'
  );
  const connectedCallbackIndex = getMethodIndex(classAST, 'connectedCallback');
  let disconnectedCallbackIndex = getMethodIndex(
    classAST,
    'disconnectedCallback'
  );

  const renderedCallback = classAST.body.body.find(
    (el) => n.MethodDefinition.check(el) && el.key.name === 'renderedCallback'
  );
  const errorCallback = classAST.body.body.find(
    (el) => n.MethodDefinition.check(el) && el.key.name === 'errorCallback'
  );

  if (connectedCallbackIndex !== -1) {
    const connectedCallback = classAST.body.body[connectedCallbackIndex];
    componentDidMount.value.body.body.push(
      ...connectedCallback.value.body.body
    );
    classAST.body.body.splice(connectedCallbackIndex, 1);
  }

  if (disconnectedCallbackIndex !== -1) {
    disconnectedCallbackIndex = getMethodIndex(
      classAST,
      'disconnectedCallback'
    );
    const disconnectedCallback = classAST.body.body[disconnectedCallbackIndex];
    componentWillUnmount.value.body.body.push(
      ...disconnectedCallback.value.body.body
    );
    classAST.body.body.splice(disconnectedCallbackIndex, 1);
  }

  if (renderedCallback) {
    renderedCallback.key = b.identifier('componentDidUpdate');
  }

  if (errorCallback) {
    errorCallback.key = b.identifier('componentDidCatch');
  }
}

function processBlock(block, options = { inConstructor: false }) {
  block.body = block.body
    .map((element) => processElement(element, options))
    .filter(Boolean);
  return block;
}

function processAssignmentExpression(assignment, options) {
  const expression = assignment.expression;

  expression.left = processExpression(expression.left, options);
  expression.right = processExpression(expression.right, options);

  return assignment;
}

function processUpdateExpression(element, options) {
  element.expression.argument = processExpression(
    element.expression.argument,
    options
  );
  return element;
}

function processReturnStatement(returnStatement, options) {
  returnStatement.argument = processExpression(
    returnStatement.argument,
    options
  );
  return returnStatement;
}

function getTopOfMemberExpression(memberExpression) {
  if (n.MemberExpression.check(memberExpression.object)) {
    return [
      ...getTopOfMemberExpression(memberExpression.object),
      memberExpression,
    ];
  } else {
    return [memberExpression];
  }
}

function processCallExpression(callExpression, options) {
  if (n.MemberExpression.check(callExpression.expression.callee)) {
    const [member] = getTopOfMemberExpression(callExpression.expression.callee);

    if (
      n.ThisExpression.check(member.object) &&
      !options.classMethods.includes(member.property.name)
    ) {
      if (member.property.name === 'template') {
        member.property.name = 'template.current';
      } else if (
        member.property.name === 'addEventListener' ||
        member.property.name === 'removeEventListener' ||
        member.property.name === 'dispatchEvent'
      ) {
        callExpression.expression.callee = b.identifier(
          'this.template.current.' + member.property.name
        );
      } else {
        callExpression.expression.callee = processExpression(
          callExpression.expression.callee,
          options
        );
      }
    }

    const lifeCycleMap = {
      renderedCallback: 'componentDidUpdate',
      connectedCallback: 'componentDidMount',
      errorCallback: 'componentDidCatch',
      disconnectedCallback: 'componentWillUnmount',
    };

    if (lifeCycleMap.hasOwnProperty(member.property.name)) {
      member.property.name = lifeCycleMap[member.property.name];
    }
  }

  callExpression.expression.arguments = callExpression.expression.arguments.map(
    (arg) => processExpression(arg, options)
  );

  return callExpression;
}

function processVariableDeclaration(element, options) {
  for (let i = 0; i < element.declarations.length; i++) {
    const declaration = element.declarations[i];

    element.declarations[i] = processExpression(
      element.declarations[i],
      options
    );

    if (n.VariableDeclarator.check(declaration)) {
      element.declarations[i] = b.variableDeclarator(
        declaration.id,
        processExpression(declaration.init, options)
      );
    }
  }
  return element;
}

function processExpression(expression, options) {
  if (n.MemberExpression.check(expression)) {
    let [top, next, ...rest] = getTopOfMemberExpression(expression);

    if (top.property.name === 'template') {
      if (next.property.name !== 'host') {
        // only add member if it is not "host"
        rest = [next, ...rest];
      }
      return b.identifier(
        `this.template.current.${rest.map((e) => e.property.name).join('.')}`
      );
    } else if (n.ThisExpression.check(expression.object)) {
      return b.identifier(`this.__s.${expression.property.name}`);
    } else if (n.MemberExpression.check(expression.object)) {
      expression.object = processExpression(expression.object, options);
      return expression;
    }
  }

  if (n.CallExpression.check(expression)) {
    return processCallExpression({ expression }, options).expression;
  }

  if (n.UpdateExpression.check(expression)) {
    expression.argument = processExpression(expression.argument, options);
    return expression;
  }

  if (n.BinaryExpression.check(expression)) {
    expression.left = processExpression(expression.left, options);
    expression.right = processExpression(expression.right, options);
    return expression;
  }

  if (n.ConditionalExpression.check(expression)) {
    expression.test = processExpression(expression.test, options);
    expression.consequent = processExpression(expression.consequent, options);
    expression.alternate = processExpression(expression.alternate, options);
  }

  if (n.ObjectExpression.check(expression)) {
    expression.properties.forEach((property) => {
      property.key = processExpression(property.key);
      property.value = processExpression(property.value);
    });
  }

  return expression;
}

function processForLoop(element, options) {
  element.test = processExpression(element.test, options);
  element.init = processVariableDeclaration(element.init, options);
  element.update = processExpression(element.update, options);
  element.body = n.BlockStatement.check(element.body)
    ? processBlock(element.body, options)
    : processExpression(element.body, options);

  return element;
}

function processForOfLoop(element, options) {
  element.left = processVariableDeclaration(element.left, options);
  element.right = processExpression(element.right, options);
  element.body = n.BlockStatement.check(element.body)
    ? processBlock(element.body, options)
    : processExpression(element.body, options);

  return element;
}

function processIfStatement(element, options) {
  element.test = processExpression(element.test, options);

  element.consequent = n.BlockStatement.check(element.consequent)
    ? processBlock(element.consequent, options)
    : processExpression(element.consequent, options);

  if (element.alternate) {
    element.alternate = n.BlockStatement.check(element.alternate)
      ? processBlock(element.alternate, options)
      : processExpression(element.alternate, options);
  }

  return element;
}

function processElement(element, options) {
  if (n.ExpressionStatement.check(element)) {
    if (n.AssignmentExpression.check(element.expression)) {
      return processAssignmentExpression(element, options);
    }
    if (n.UpdateExpression.check(element.expression)) {
      return processUpdateExpression(element, options);
    }

    if (n.CallExpression.check(element.expression)) {
      return processCallExpression(element, options);
    }
  } else if (n.VariableDeclaration.check(element)) {
    return processVariableDeclaration(element, options);
  } else if (n.IfStatement.check(element)) {
    return processIfStatement(element, options);
  } else if (n.ReturnStatement.check(element)) {
    return processReturnStatement(element, options);
  } else if (n.ForStatement.check(element)) {
    return processForLoop(element, options);
  } else if (n.ForOfStatement.check(element)) {
    return processForOfLoop(element, options);
  }

  return element;
}

function buildStyles(classAST, templateIdentifier) {
  const classBuilder = parse(`
    class A{
      componentDidMount() {
        super.componentDidMount(${templateIdentifier});
      }
      componentWillUnmount() {
        super.componentWillUnmount(${templateIdentifier});
      }
    }
  `);

  classAST.body.body.push(
    classBuilder.program.body[0].body.body[0] // componentDidMount
  );

  classAST.body.body.push(
    classBuilder.program.body[0].body.body[1] // componentWillUnmount
  );
}

function convertMethods(classAST, publicProps, classMethods) {
  return classAST.body.body
    .filter(
      (property) =>
        property?.kind === 'method' ||
        property?.kind === 'get' ||
        property?.kind === 'set'
    )
    .map((method) => {
      return b.methodDefinition(
        method.kind,
        b.identifier(method.key.name),
        convertMethod(method.value, {
          inConstructor: false,
          publicProps,
          classMethods,
        })
      );
    });
}

function convertMethod(method, options) {
  method.body.body = method.body.body
    .map((element) => processElement(element, options))
    .filter(Boolean);

  return method;
}

export function compileClass(ast) {
  const classAST = getClassAST(ast);

  if (classAST) {
    const publicProps = getPublicProps(ast);

    removeLWCCode(ast);

    ast.program.body.unshift(
      parse("import ObservableMembrane from 'observable-membrane'").program
        .body[0]
    );

    ast.program.body.unshift(
      parse("import LightningComponent from 'lwc2react/lib/LightningComponent'")
        .program.body[0]
    );

    const exportStatement = ast.program.body.find((el) =>
      n.ExportDefaultDeclaration.check(el)
    );
    const templateIdentifier = exportStatement.declaration.arguments[1].properties.find(
      (prop) => prop.key.name === 'tmpl'
    ).value.name;

    exportStatement.declaration = exportStatement.declaration.arguments[0];

    convertClass(classAST, publicProps, templateIdentifier);
  }

  return prettyPrint(ast, { tabWidth: 2 }).code;
}
