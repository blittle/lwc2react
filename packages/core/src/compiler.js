import { parse } from 'recast';
import { compileClass } from './class-compiler';
import { compileTemplate } from './template-compiler';

export function compile(id, source) {
  if (id.includes('@lwc/engine/dist/engine.js') || id.includes('wire-service'))
    return 'export default undefined';

  const ast = parse(source);

  if (id.endsWith('.html')) {
    return compileTemplate(ast);
  } else if (id.endsWith('.css')) {
    return source;
  } else {
    return compileClass(ast);
  }
}
