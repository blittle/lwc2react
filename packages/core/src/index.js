import pluginUtils from '@rollup/pluginutils';
import { compile } from './compiler';

export default function rollupLWC2ReactCompiler(pluginOptions = {}) {
  let { include, exclude, debug } = pluginOptions;

  const filter = pluginUtils.createFilter(include, exclude);

  return {
    name: 'rollup-lwc2react-compiler',
    async transform(src, id) {
      if (debug && filter(id)) {
        console.log('lwc2react -> ', id);
      }

      if (!filter(id)) {
        return;
      }

      let code;
      try {
        code = compile(id, src);
        if (code) {
          if (debug && id.includes(debug)) {
            console.log(src);
            console.log('===========');
            console.log(code);
          }
          return {
            code,
            map: null,
          };
        }
      } catch (error) {
        console.log('error processing lwc source: ', id);
        console.log(src);
        throw error;
      }
    },
  };
}
