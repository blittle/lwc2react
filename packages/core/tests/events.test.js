import { compile } from '../src/compiler';

describe('Events', function () {
  it('bound events in template', function () {
    const source = `
import _implicitStylesheets from "./app.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    d: api_dynamic,
    h: api_element,
    t: api_text,
    b: api_bind
  } = $api;
  const {
    _m0
  } = $ctx;
  return [api_element("div", {
    key: 2
  }, [api_element("span", {
    key: 0
  }, [api_dynamic($cmp.something)]), api_element("button", {
    key: 1,
    on: {
      "click": _m0 || ($ctx._m0 = api_bind($cmp.increment))
    }
  }, [api_text("increment")])])];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-app_app-host",
  shadowAttribute: "my-app_app"
};
    `;

    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should render input form events and attributes', function () {
    const source = `
import _implicitStylesheets from "./searchBar.css";
import { registerTemplate } from "lwc";
function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    h: api_element,
    b: api_bind
  } = $api;
  const {
    _m0,
    _m1,
    _m2
  } = $ctx;
  return [api_element("div", {
    classMap: {
      "site-search": true
    },
    key: 3
  }, [api_element("span", {
    classMap: {
      "fa": true,
      "fa-search": true
    },
    key: 0
  }, []), api_element("input", {
    classMap: {
      "form-control": true,
      "search-field": true
    },
    attrs: {
      "placeholder": "Search (keywords, brands, etc)",
      "type": "search"
    },
    props: {
      "value": $cmp.query,
      "autofocus": true
    },
    key: 1,
    on: {
      "input": _m0 || ($ctx._m0 = api_bind($cmp.searchInputHandler)),
      "keyup": _m1 || ($ctx._m1 = api_bind($cmp.searchKeyUpHandler))
    }
  }, []), api_element("i", {
    classMap: {
      "fa": true,
      "fa-search": true
    },
    key: 2,
    on: {
      "click": _m2 || ($ctx._m2 = api_bind($cmp.handleIconClick))
    }
  }, [])])];
}
export default registerTemplate(tmpl);
tmpl.stylesheets = [];
if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-searchBar_searchBar-host",
  shadowAttribute: "my-searchBar_searchBar"
};`;

    expect(compile('some.html', source)).toMatchSnapshot();
  });

  it('should dispatch custom events from the template', function () {
    const source = `
import { registerDecorators as _registerDecorators } from "lwc";
import _tmpl from "./searchBar.html";
import { registerComponent as _registerComponent } from "lwc";
import { LightningElement } from 'lwc';
/**
 * Search Bar where visitors can search for stuff
 */
class SearchBar extends LightningElement {
  constructor(...args) {
    super(...args);
    this.query = '';
  }

  performSearch() {
    this.template.dispatchEvent(new CustomEvent('someevent', {
      detail: this.query
    }));
  }
}
_registerDecorators(SearchBar, {
  publicProps: {
    query: {
      config: 0
    }
  }
})
export default _registerComponent(SearchBar, {
  tmpl: _tmpl
});
    `;
    expect(compile('something.js', source)).toMatchSnapshot();
  });

  it('should dispatch custom events from the component ref', function () {
    const source = `
import { registerDecorators as _registerDecorators } from "lwc";
import _tmpl from "./searchBar.html";
import { registerComponent as _registerComponent } from "lwc";
import { LightningElement } from 'lwc';
/**
 * Search Bar where visitors can search for stuff
 */
class SearchBar extends LightningElement {
  constructor(...args) {
    super(...args);
    this.query = '';
  }

  performSearch() {
    this.dispatchEvent(new CustomEvent('someevent', {
      detail: this.query
    }));
  }
}
_registerDecorators(SearchBar, {
  publicProps: {
    query: {
      config: 0
    }
  }
})
export default _registerComponent(SearchBar, {
  tmpl: _tmpl
});
    `;
    expect(compile('something.js', source)).toMatchSnapshot();
  });

  it('should add event listeners to ref', function () {
    const source = `
import { registerDecorators as _registerDecorators } from "lwc";
import _tmpl from "./searchBar.html";
import { registerComponent as _registerComponent } from "lwc";
import { LightningElement } from 'lwc';
/**
 * Search Bar where visitors can search for stuff
 */
class SearchBar extends LightningElement {
  constructor(...args) {
    super(...args);
    this.query = '';
    this.template.addEventListener('something', this.performSearch);
    this.addEventListener('something', this.performSearch);
  }

  performSearch() {}
}
_registerDecorators(SearchBar, {
  publicProps: {
    query: {
      config: 0
    }
  }
})
export default _registerComponent(SearchBar, {
  tmpl: _tmpl
});
    `;
    expect(compile('something.js', source)).toMatchSnapshot();
  });

  it('should convert custom events on templates', function () {
    const source = `
import _implicitStylesheets from "./header.css";
import _mySearchBar from "my/searchBar";
import { registerTemplate } from "lwc";
function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    b: api_bind,
    c: api_custom_element
  } = $api;
  const {
    _m0
  } = $ctx;
  return [api_custom_element("my-search-bar", _mySearchBar, {
    props: {
      "query": $cmp.query
    },
    key: 0,
    on: {
      "search": _m0 || ($ctx._m0 = api_bind($cmp.updateQuery))
    }
  }, [])];
}
export default registerTemplate(tmpl);
tmpl.stylesheets = [];
if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-header_header-host",
  shadowAttribute: "my-header_header"
};
    `;

    expect(compile('something.html', source)).toMatchSnapshot();
  });
});
