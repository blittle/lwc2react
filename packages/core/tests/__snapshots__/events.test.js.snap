// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events bound events in template 1`] = `
"import _implicitStylesheets from \\"./app.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 2,
    ref: $cmp.template
  }, [React.createElement(\\"span\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0
  }, $cmp.something), React.createElement(\\"button\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 1,
    onClick: $cmp.increment.bind($cmp)
  }, \\"increment\\")]);
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-app_app-host\\",
  shadowAttribute: \\"my-app_app\\"
};

tmpl.customEvents = [];"
`;

exports[`Events should add event listeners to ref 1`] = `
"import ObservableMembrane from \\"observable-membrane\\";
import React from \\"react\\";
import _tmpl from \\"./searchBar.html\\";

/**
 * Search Bar where visitors can search for stuff
 */
class SearchBar extends React.Component {
  constructor(...args) {
    super(...args);

    const membrane = new ObservableMembrane({
      valueMutated: () => {
        if (this.mounted)
          this.forceUpdate();
      }
    });

    this.__s = membrane.getProxy({
      performSearch: this.performSearch
    });

    this.template = React.createRef();
    this.__s.query = \\"\\";
    this.template.addEventListener(\\"something\\", this.__s.performSearch);
    this.template.current.addEventListener(\\"something\\", this.__s.performSearch);
  }

  performSearch() {}

  render() {
    return _tmpl(Object.assign(this, this.props, this.__s));
  }

  componentDidMount() {
    this.mounted = true;
    this.stylesheets = [];

    _tmpl.stylesheets.forEach(stylesheet => {
      const sheet = document.createElement(\\"style\\");
      sheet.type = \\"text/css\\";

      sheet.textContent = stylesheet(
        \\"[\\" + _tmpl.stylesheetTokens.hostAttribute.toLowerCase() + \\"]\\",
        \\"[\\" + _tmpl.stylesheetTokens.shadowAttribute.toLowerCase() + \\"]\\",
        null
      );

      document.head.appendChild(sheet);
      this.stylesheets.push(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this[ref] = this[ref].bind(this);
      this.template.current.addEventListener(name, this[ref]);
    });
  }

  componentWillUnmount() {
    this.mounted = false;

    this.stylesheets.forEach(sheet => {
      if (sheet.parentNode)
        sheet.parentNode.removeChild(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this.template.current.removeEventListener(name, this[ref]);
    });
  }
}

export default SearchBar;"
`;

exports[`Events should convert custom events on templates 1`] = `
"import _implicitStylesheets from \\"./header.css\\";
import _mySearchBar from \\"my/searchBar\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(_mySearchBar, {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    ref: $cmp.template,
    \\"query\\": $cmp.query
  }, null);
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-header_header-host\\",
  shadowAttribute: \\"my-header_header\\"
};

tmpl.customEvents = [[\\"search\\", \\"updateQuery\\"]];"
`;

exports[`Events should dispatch custom events from the component ref 1`] = `
"import ObservableMembrane from \\"observable-membrane\\";
import React from \\"react\\";
import _tmpl from \\"./searchBar.html\\";

/**
 * Search Bar where visitors can search for stuff
 */
class SearchBar extends React.Component {
  constructor(...args) {
    super(...args);

    const membrane = new ObservableMembrane({
      valueMutated: () => {
        if (this.mounted)
          this.forceUpdate();
      }
    });

    this.__s = membrane.getProxy({
      performSearch: this.performSearch
    });

    this.template = React.createRef();
    this.__s.query = \\"\\";
  }

  performSearch() {
    this.template.current.dispatchEvent(new CustomEvent(\\"someevent\\", {
      detail: this.query
    }));
  }

  render() {
    return _tmpl(Object.assign(this, this.props, this.__s));
  }

  componentDidMount() {
    this.mounted = true;
    this.stylesheets = [];

    _tmpl.stylesheets.forEach(stylesheet => {
      const sheet = document.createElement(\\"style\\");
      sheet.type = \\"text/css\\";

      sheet.textContent = stylesheet(
        \\"[\\" + _tmpl.stylesheetTokens.hostAttribute.toLowerCase() + \\"]\\",
        \\"[\\" + _tmpl.stylesheetTokens.shadowAttribute.toLowerCase() + \\"]\\",
        null
      );

      document.head.appendChild(sheet);
      this.stylesheets.push(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this[ref] = this[ref].bind(this);
      this.template.current.addEventListener(name, this[ref]);
    });
  }

  componentWillUnmount() {
    this.mounted = false;

    this.stylesheets.forEach(sheet => {
      if (sheet.parentNode)
        sheet.parentNode.removeChild(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this.template.current.removeEventListener(name, this[ref]);
    });
  }
}

export default SearchBar;"
`;

exports[`Events should dispatch custom events from the template 1`] = `
"import ObservableMembrane from \\"observable-membrane\\";
import React from \\"react\\";
import _tmpl from \\"./searchBar.html\\";

/**
 * Search Bar where visitors can search for stuff
 */
class SearchBar extends React.Component {
  constructor(...args) {
    super(...args);

    const membrane = new ObservableMembrane({
      valueMutated: () => {
        if (this.mounted)
          this.forceUpdate();
      }
    });

    this.__s = membrane.getProxy({
      performSearch: this.performSearch
    });

    this.template = React.createRef();
    this.__s.query = \\"\\";
  }

  performSearch() {
    this.template.dispatchEvent(new CustomEvent(\\"someevent\\", {
      detail: this.query
    }));
  }

  render() {
    return _tmpl(Object.assign(this, this.props, this.__s));
  }

  componentDidMount() {
    this.mounted = true;
    this.stylesheets = [];

    _tmpl.stylesheets.forEach(stylesheet => {
      const sheet = document.createElement(\\"style\\");
      sheet.type = \\"text/css\\";

      sheet.textContent = stylesheet(
        \\"[\\" + _tmpl.stylesheetTokens.hostAttribute.toLowerCase() + \\"]\\",
        \\"[\\" + _tmpl.stylesheetTokens.shadowAttribute.toLowerCase() + \\"]\\",
        null
      );

      document.head.appendChild(sheet);
      this.stylesheets.push(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this[ref] = this[ref].bind(this);
      this.template.current.addEventListener(name, this[ref]);
    });
  }

  componentWillUnmount() {
    this.mounted = false;

    this.stylesheets.forEach(sheet => {
      if (sheet.parentNode)
        sheet.parentNode.removeChild(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this.template.current.removeEventListener(name, this[ref]);
    });
  }
}

export default SearchBar;"
`;

exports[`Events should render input form events and attributes 1`] = `
"import _implicitStylesheets from \\"./searchBar.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 3,
    ref: $cmp.template,
    className: \\"site-search\\"
  }, [React.createElement(\\"span\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    className: \\"fa fa-search\\"
  }, null), React.createElement(\\"input\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 1,
    onInput: $cmp.searchInputHandler.bind($cmp),
    onKeyUp: $cmp.searchKeyUpHandler.bind($cmp),
    \\"value\\": $cmp.query,
    \\"autoFocus\\": true,
    \\"placeholder\\": \\"Search (keywords, brands, etc)\\",
    \\"type\\": \\"search\\",
    className: \\"form-control search-field\\"
  }, null), React.createElement(\\"i\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 2,
    onClick: $cmp.handleIconClick.bind($cmp),
    className: \\"fa fa-search\\"
  }, null)]);
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-searchBar_searchBar-host\\",
  shadowAttribute: \\"my-searchBar_searchBar\\"
};

tmpl.customEvents = [];"
`;
